### PR TITLE
feat: display notification badge on user account image - WPB-11270

### PR DESCRIPTION
**Pull request by El-Fitz**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-11270" title="WPB-11270" target="_blank">StoryWPB-11270</a>  [iOS] Start and follow team creation flow for personal users
  </td></table>
  <br />
 

### Issue

Display a notification badge on user account image when they aren't part of a team, API v7 is available, and they haven't yet seen their profile. Hide it once they've viewed their profile.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift [hunk 1]
```diff
@@ -314,7 +337,7 @@ struct AccountImageView_Previews: PreviewProvider {
         _ availability: Availability?
     ) -> some View {
         NavigationStack {
-            AccountImageViewRepresentable(source, availability)
+            AccountImageViewRepresentable(source, availability, .many)
```
Comment (netbe) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift]: what's .many?

Comment (netbe) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift]: why don't we have named parameter here

File: WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import SwiftUI
+import WireDesign
+
+// MARK: -
+
+final class NotificationBadgeView: UIView {
```
Comment (netbe) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift]: I wonder if that could have been done quicker in SwiftUI

File: WireUI/Sources/WireDesign/Icons/UImage+ImageResource.swift [hunk 3]
```diff
@@ -1,6 +1,6 @@
 //
 // Wire
-// Copyright (C) 2024 Wire Swiss GmbH
+// Copyright (C) 2024 Wire Swi#ss GmbH
```
Comment (netbe) [inline WireUI/Sources/WireDesign/Icons/UImage+ImageResource.swift]:
```suggestion
// Copyright (C) 2024 Wire Swiss GmbH
```

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift [hunk 4]
```diff
@@ -210,6 +241,19 @@ extension ConversationListViewController.ViewModel {
...
         }
+
+        userDidViewSelfProfileToken = notificationCenter.addObserver(
+            forName: .userDidViewSelfProfile,
+            object: nil,
+            queue: .main
+        ) { [weak self] _ in
+            guard let self else { return }
+            let userDefaults = PrivateUserDefaults<UserDefaultsKey>(userID: userSession.selfUser.remoteIdentifier)
+            userDefaults.set(true, forKey: .didViewSelfProfile)
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: maybe hide this under a method of the viewModel

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift [hunk 5]
```diff
@@ -115,6 +124,25 @@ extension ConversationListViewController {
...
+            if userSession.selfUser.isTeamMember {
+                return nil
+            }
+            guard let apiVersion = BackendInfo.apiVersion,
+                  apiVersion > .v6 else {
+                return nil
+            }
+
+            let userDefaults = PrivateUserDefaults<UserDefaultsKey>(userID: userSession.selfUser.remoteIdentifier)
+            let value = userDefaults.object(forKey: .didViewSelfProfile)
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: could be hidden under a convernience method:

var didViewSelfProfile: Bool {
get set...
}

**Review by netbe — COMMENTED** [review]
early review, left a couple comments

File: WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift (hunk 2 above)
Comment (netbe) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift]: but for now don't change it

File: WireUI/Sources/WireDesign/Icons/Image+ImageResource.swift [hunk 6]
```diff
@@ -21,4 +21,5 @@ import SwiftUI
 public extension Image {
     static let close = Image(.close)
     static let info = Image(.info)
+    static let infoCircle = Image(.infoCircle)
```
Comment (caldrian) [inline WireUI/Sources/WireDesign/Icons/Image+ImageResource.swift]: I strongly encourage to not have feature specific image assets at a central point.
Please move it to the target where it is used.
If this should be a view component to be reused, please move it to `WireReusableUIComponents`.

File: WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift [hunk 7]
```diff
@@ -93,12 +93,16 @@ public final class AccountImageView: UIView {
         set { availabilityIndicatorView.backgroundViewColor = newValue }
     }
 
+    public var notifications: ProfileNotifications? {
```
Comment (caldrian) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift]: Why not just a simple Bool like `isNotificationBadgeVisible`?

File: WireUI/Sources/WireAccountImageUI/AccountImageView/ProfileNotifications.swift [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+public enum ProfileNotifications {
+    case many
+}
```
Comment (caldrian) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/ProfileNotifications.swift]: It really doesn't make any sense to me to define a custom type with a single value and then use it as optional, to add a second value.
We should use a boolean instead.

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift (hunk 4 above)
Comment (netbe) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: done

File: WireUI/Sources/WireDesign/Icons/Image+ImageResource.swift (hunk 6 above)
Comment (netbe) [reply WireUI/Sources/WireDesign/Icons/Image+ImageResource.swift]: done

File: WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift (hunk 7 above)
Comment (netbe) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift]: replaced

File: WireUI/Sources/WireAccountImageUI/AccountImageView/ProfileNotifications.swift (hunk 8 above)
Comment (netbe) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/ProfileNotifications.swift]: done

**Review by netbe — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift [hunk 9]
```diff
@@ -115,6 +124,36 @@ extension ConversationListViewController {
...
+        var didViewSelfProfile: Bool {
+            get {
+                let userDefaults = PrivateUserDefaults<UserDefaultsKey>(userID: userSession.selfUser.remoteIdentifier)
+                let value = userDefaults.object(forKey: .didViewSelfProfile)
+
+                if value == nil {
+                    userDefaults.set(false, forKey: .didViewSelfProfile)
+                }
+
+                return (value as? Bool) == true
```
Comment (El-Fitz) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: This doesn't work when the value has never been set

File: WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift (hunk 2 above)
Comment (El-Fitz) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift]: No. I just reused and simplified the code from the availability view.
What took time is integrating it and properly updating its state.

**Comment by El-Fitz** [comment]
Updated it for the sidebar.
Also extracted the flag's storage so we can be sure it's updated in the UserDefaults before the notification is sent to observers, and to avoid duplicating the logic and user defaults key.

https://github.com/user-attachments/assets/68d05f35-5ca1-4dba-a66a-26df780a1f7c


https://github.com/user-attachments/assets/2a082956-d69e-4ace-9dcb-c2a1a674f2f6

Simulator Screenshot - iPhone 16 - 2024-12-12 at 08 48 12

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift [hunk 10]
```diff
@@ -1082,6 +1082,7 @@ public final class SessionManager: NSObject, SessionManagerType {
         deleteTemporaryData()
 
         PrivateUserDefaults.removeAll(forUserID: account.userIdentifier, in: sharedUserDefaults)
+        PrivateUserDefaults.removeAll(forUserID: account.userIdentifier, in: .standard)
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:1085]: Why is this change needed?

**Review by caldrian — COMMENTED** [review]
Please add at least one snapshot test where the notification dot can be seen.

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift (hunk 5 above)
Comment (El-Fitz) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: Moved into a separate object

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift (hunk 4 above)
Comment (El-Fitz) [reply wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: Moved into a separate object

**Review by caldrian — COMMENTED** [review]
Screenshot 2024-12-12 at 12 03 44

There is some duplicate white border.
I think the inner one needs to go, it should be red as well.

File: WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift [hunk 11]
```diff
@@ -232,9 +236,25 @@ public final class AccountImageView: UIView {
...
+        NSLayoutConstraint.activate([
+            notificationBadgeView.widthAnchor.constraint(
+                equalToConstant: 14
+            ),
+            notificationBadgeView.heightAnchor.constraint(equalToConstant: 14),
+            notificationBadgeView.centerYAnchor.constraint(equalTo: accountImageViewWrapper.topAnchor, constant: 3.5),
+            notificationBadgeView.centerXAnchor.constraint(
+                equalTo: accountImageViewWrapper.trailingAnchor,
+                constant: -3.5
+            )
```
Comment (caldrian) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift:252]: Could you try to make the size of the dot relative, please?
I think the remaining code somehow scales properly. (The view is not always displayed in the same size, like between iPhone and iPad [sidebar])
ezgif-4-a979cb42f5

Comment (caldrian) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift:252]: Actually the border around the image doesn't seem to scale, but still, it would be nice.

File: WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift [hunk 12]
```diff
@@ -232,9 +236,25 @@ public final class AccountImageView: UIView {
             accountImageViewWrapper.bottomAnchor.constraint(equalTo: availabilityIndicatorView.bottomAnchor)
         ])
 
+        // view which renders the notification badge
+        notificationBadgeView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(notificationBadgeView)
```
Comment (caldrian) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/AccountImageView.swift:242]: Can't we just use a simple `UIImageView`?

File: WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        didSet { setNeedsLayout() }
+    }
+
+    // MARK: - Private Properties
+
+    /// A view which serves as background and outer border.
+    private let backgroundView = UIView()
+
+    /// An image view for the icon
+    private let iconView = UIImageView(image: .info)
```
Comment (caldrian) [inline WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift]: not urgent now, but in my opinion we should get rid of the extensions to `UIImage` and `Image`
```
public extension UIImage {
    static let infoCircle = UIImage(resource: .infoCircle)
}
```
because they almost don't add any value:
```suggestion
    private let iconView = UIImageView(image: .init(resource: .info))
```

Comment (caldrian) [reply WireUI/Sources/WireAccountImageUI/AccountImageView/NotificationBadgeView.swift]: Even worse for the SwiftUI variant:
```
let infoCircle = Image.infoCircle
```
vs
```
let infoCircle = Image(.infoCircle)
```

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift [hunk 14]
```diff
@@ -115,6 +120,18 @@ extension ConversationListViewController {
...
+        public var showProfileNotificationsBadge: Bool {
+            if userSession.selfUser.isTeamMember {
+                return false
+            }
+            guard let apiVersion = BackendInfo.apiVersion,
+                  apiVersion >= .v7 else {
+                return false
+            }
+
+            return !SelfProfileViewsMonitor.shared.didViewSelfProfile
```
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ViewModel/ConversationListViewControllerViewModel.swift]: Please don't use singletons.

File: wire-ios/Wire-iOS/Sources/UserInterface/MainController/ZClientViewController.swift [hunk 15]
```diff
@@ -772,10 +774,12 @@ final class ZClientViewController: UIViewController {
...
     private func updateCachedAccountInfo() async {
         do {
+            let user = userSession.selfUser
             cachedAccountInfo = SidebarAccountInfo(
-                userSession.selfUser,
+                user,
                 cachedAccountImage,
-                cachedAccountInfo.isE2EICertified
+                cachedAccountInfo.isE2EICertified,
+                showNotificationsBadge: shouldShowNotificationsBadge(user: user)
```
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/MainController/ZClientViewController.swift:783]: optional
```suggestion
                isNotificationsBadgeVisible: ...
```

File: wire-ios/Wire-iOS/Sources/UserInterface/SelfProfile/SelfProfileViewController.swift [hunk 16]
```diff
@@ -162,6 +162,7 @@ final class SelfProfileViewController: UIViewController {
 
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
+        SelfProfileViewsMonitor.shared.onDidViewSelfProfile()
```
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/SelfProfile/SelfProfileViewController.swift]: Please don't use singletons.
Ideally we want to test the `SelfProfileViewController` individually without side effects.
For that we usually declare protocols for any dependencies.
We also use Sourcery for auto-creating the mocks.

File: wire-ios/Wire-iOS/Sources/UserInterface/MainController/SidebarViewControllerBuilder.swift [hunk 17]
```diff
@@ -34,10 +34,12 @@ struct SidebarViewControllerBuilder {
         let legalHoldIndicatorViewDesign = LegalHoldIndicatorViewDesign()
 
         let sidebarViewController = SidebarViewController(
-            accountImageView: { accountImage, availability in
+            accountImageView: { accountImage, availability, showNotificationsBadge in
                 AccountImageViewRepresentable(
                     source: accountImage.mapToAccountImageSource(),
-                    availability: availability?.mapToAccountImageAvailability()
+                    availability: availability?.mapToAccountImageAvailability(),
+                    // TODO: [WPB-11968] Use the actual value based on team membership, available API version & UserDefaults,
```
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/UserInterface/MainController/SidebarViewControllerBuilder.swift]: Ticket WPB-11968 is marked as in code review. Is there another PR which solves this TODO?

Comment (El-Fitz) [reply wire-ios/Wire-iOS/Sources/UserInterface/MainController/SidebarViewControllerBuilder.swift]: The comment is outdated.

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift (hunk 10 above)
Comment (netbe) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:1085]: I remove the user default storage which includes the notification badge on logout.

**Review by netbe — APPROVED** [review]
Looks good to me,
@El-Fitz  I updated the tests to add the showNotifications snapshots

**Review by caldrian — APPROVED** [review]
I think the dark mode icon is missing for the notification dot.
Maybe leave a TODO somewhere.
Good work!
